libs/wire-api/src/Wire/API/Routes/Version.hs

```diff
@@ -57,44 +57,68 @@ import Servant
 import Servant.Swagger
 import Wire.API.Routes.Named
 import Wire.API.VersionInfo
+import Wire.Arbitrary (Arbitrary, GenericUniform (GenericUniform))
 
--- | Version of the public API.
+-- | Version of the public API.  Serializes to `"v<n>"`.  See 'VersionNumber' below for one
+-- that serializes to `<n>`.  See `/libs/wire-api/test/unit/Test/Wire/API/Routes/Version.hs`
+-- for serialization rules.
 data Version = V0 | V1 | V2 | V3
-  deriving stock (Eq, Ord, Bounded, Enum, Show)
+  deriving stock (Eq, Ord, Bounded, Enum, Show, Generic)
   deriving (FromJSON, ToJSON) via (Schema Version)
+  deriving (Arbitrary) via (GenericUniform Version)
+
+versionString :: IsString a => Version -> a
+versionString V0 = "v0"
+versionString V1 = "v1"
+versionString V2 = "v2"
+versionString V3 = "v3"
+
+versionInt :: Integral i => Version -> i
+versionInt V0 = 0
+versionInt V1 = 1
+versionInt V2 = 2
+versionInt V3 = 3
 
 instance ToSchema Version where
-  schema =
-    enum @Integer "Version" . mconcat $
-      [ element 0 V0,
-        element 1 V1,
-        element 2 V2,
-        element 3 V3
-      ]
-
-mkVersion :: Integer -> Maybe Version
-mkVersion n = case Aeson.fromJSON (Aeson.Number (fromIntegral n)) of
-  Aeson.Error _ -> Nothing
-  Aeson.Success v -> pure v
+  schema = enum @Text "Version" . mconcat $ (\v -> element (versionString v) v) <$> [minBound ..]
 
 instance FromHttpApiData Version where
+  parseQueryParam v = note ("Unknown version: " <> v) $
+    getAlt $
+      flip foldMap [minBound ..] $ \s ->
+        guard (versionString s == v) $> s
+
+instance ToHttpApiData Version where
+  toHeader = versionString
+  toUrlPiece = versionString
+
+instance ToByteString Version where
+  builder = versionString
+
+-- | Wrapper around 'Version' that serializes to integers `<n>`, as needed in
+-- eg. `VersionInfo`.  See `/libs/wire-api/test/unit/Test/Wire/API/Routes/Version.hs` for
+-- serialization rules.
+newtype VersionNumber = VersionNumber {fromVersionNumber :: Version}
+  deriving stock (Eq, Ord, Show, Generic)
+  deriving newtype (Bounded, Enum)
+  deriving (FromJSON, ToJSON) via (Schema VersionNumber)
+  deriving (Arbitrary) via (GenericUniform Version)
+
+instance ToSchema VersionNumber where
+  schema =
+    enum @Integer "Version" . mconcat $ (\v -> element (versionInt v) (VersionNumber v)) <$> [minBound ..]
```

Comment:
```suggestion
    enum @Integer "VersionNumber" . mconcat $ (\v -> element (versionInt v) (VersionNumber v)) <$> [minBound ..]
```

Have you chosen the description string intentionally or did this happen by copy & paste? (If you had a reason to use `"Version"` here, please ignore this comment.)